### PR TITLE
Block execution invariant

### PR DESCRIPTION
**Pull request by SebastianMarian**
* Implemented block execution invariant as a final check after a block is created or processed (this should avoid situation when a cross shard miniblock would generate a stuck situation in destination shard due to gas limit exceed generated by an old/new bug introduced. Also it should avoid situation when an overflowed amount of accumulated/developer fees are generated due some existing or newly bugs introduced)

### CODE REVIEW

File: process/coordinator/process.go [hunk 1]
```diff
@@ -46,14 +47,18 @@ type transactionCoordinator struct {
...
+	feeHandler                        process.TransactionFeeHandler
+	blockSizeComputation              preprocess.BlockSizeComputationHandler
+	balanceComputation                preprocess.BalanceComputationHandler
+	requestedItemsHandler             process.TimeCacher
+	economicsFee                      process.FeeHandler
+	txTypeHandler                     process.TxTypeHandler
+	blockGasAndFeesReCheckEnableEpoch uint32
 }
 
+//TODO: Should be refactored with arguments (added task EN-8790 in Jira)
```
Comment (iulianpascalau) [inline process/coordinator/process.go:61]: :+1:

File: process/coordinator/process.go [hunk 2]
```diff
@@ -1162,6 +1179,147 @@ func getNumOfCrossShardScCallsOrSpecialTxs(
 	return numCrossShardScCallsOrSpecialTxs
 }
 
+// VerifyCreatedMiniBlocks re-checks gas used and generated fees in the given block
+func (tc *transactionCoordinator) VerifyCreatedMiniBlocks(header data.HeaderHandler, body *block.Body) error {
+	if header.GetEpoch() < tc.blockGasAndFeesReCheckEnableEpoch {
+		return nil
+	}
+
+	mapMiniBlockTypeAllTxs := make(map[block.Type]map[string]data.TransactionHandler)
```
Comment (iulianpascalau) [inline process/coordinator/process.go:1188]: if the lines L1188-L1194 would have been extracted in a small function, the VerifyCreatedMiniBlocks would have had a perfect structure as all its contents would have been on the same abstraction level.

File: process/coordinator/process.go [hunk 3]
```diff
@@ -1162,6 +1179,147 @@ func getNumOfCrossShardScCallsOrSpecialTxs(
...
+		if !ok {
+			log.Debug("missing transaction in checkGasConsumedByMiniBlockInReceiverShard ", "type", miniBlock.Type, "txHash", txHash)
+			return process.ErrMissingTransaction
+		}
+
+		_, txTypeDstShard := tc.txTypeHandler.ComputeTransactionType(txHandler)
+		if txTypeDstShard == process.MoveBalance {
+			gasConsumedByTxInReceiverShard = tc.economicsFee.ComputeGasLimit(txHandler)
+		} else {
+			gasConsumedByTxInReceiverShard = txHandler.GetGasLimit()
```
Comment (iulianpascalau) [inline process/coordinator/process.go:1248]: Is this correct?
Shouldn't this be txHandler.GetGasLimit() - gas in sender shard?

File: process/coordinator/process.go [hunk 4]
```diff
@@ -1162,6 +1179,147 @@ func getNumOfCrossShardScCallsOrSpecialTxs(
...
+		txHandler, ok := mapHashTx[string(txHash)]
+		if !ok {
+			log.Debug("missing transaction in getMaxAccumulatedFeesAndDeveloperFees ", "type", miniBlock.Type, "txHash", txHash)
+			return big.NewInt(0), big.NewInt(0), process.ErrMissingTransaction
+		}
+
+		maxAccumulatedFeesFromTx := core.SafeMul(txHandler.GetGasLimit(), txHandler.GetGasPrice())
+		maxAccumulatedFeesFromMiniBlock.Add(maxAccumulatedFeesFromMiniBlock, maxAccumulatedFeesFromTx)
+
+		maxDeveloperFeesFromTx := core.GetPercentageOfValue(maxAccumulatedFeesFromTx, tc.economicsFee.DeveloperPercentage())
```
Comment (iulianpascalau) [inline process/coordinator/process.go:1316]: Heads up, this function is changed in https://github.com/ElrondNetwork/elrond-go/pull/2760

File: process/coordinator/process_test.go [hunk 5]
```diff
@@ -2813,3 +3016,6 @@ func TestTransactionCoordinator_GetNumOfCrossShardSpecialTxsShouldWork(t *testin
 	}
 	assert.Equal(t, 1, getNumOfCrossShardScCallsOrSpecialTxs(mb, allTxs, 0))
 }
+
+//TODO: Unit tests for methods: VerifyCreatedMiniBlocks, verifyGasLimit, checkGasConsumedByMiniBlockInReceiverShard,
```
Comment (iulianpascalau) [inline process/coordinator/process_test.go:3020]: Maybe add a Jira task so we won't forget about this?

Comment (SebastianMarian) [reply process/coordinator/process_test.go:3020]: I will add them also in the current task anyway, in the next commit, but first I added only the productive code just to be easily to be reviewed

File: process/coordinator/process.go (hunk 4 above)
Comment (SebastianMarian) [reply process/coordinator/process.go:1316]: I will check it after #2760 merge :+1:

File: process/coordinator/process.go (hunk 3 above)
Comment (SebastianMarian) [reply process/coordinator/process.go:1248]: Yep, changed this way :+1:

File: process/coordinator/process.go (hunk 2 above)
Comment (SebastianMarian) [reply process/coordinator/process.go:1188]: Done

**Review by valentincraciun99 — DISMISSED** [review]
system testing

**Review by iulianpascalau — APPROVED** [review]

File: process/coordinator/process.go [hunk 6]
```diff
@@ -1162,6 +1179,157 @@ func getNumOfCrossShardScCallsOrSpecialTxs(
...
+		return err
+	}
+
+	return nil
+}
+
+func (tc *transactionCoordinator) getAllTransactions(body *block.Body) map[block.Type]map[string]data.TransactionHandler {
+	mapMiniBlockTypeAllTxs := make(map[block.Type]map[string]data.TransactionHandler)
+	for _, miniBlock := range body.MiniBlocks {
+		_, ok := mapMiniBlockTypeAllTxs[miniBlock.Type]
```
Comment (raduchis) [inline process/coordinator/process.go:1206]: could there be 2 miniblocks with the same type? Will the second miniblock of the same type be taken into consideration?

**Review by iulianpascalau — APPROVED** [review]

File: process/coordinator/process.go [hunk 7]
```diff
@@ -1162,6 +1179,157 @@ func getNumOfCrossShardScCallsOrSpecialTxs(
...
+}
+
+func (tc *transactionCoordinator) verifyGasLimit(
+	body *block.Body,
+	mapMiniBlockTypeAllTxs map[block.Type]map[string]data.TransactionHandler,
+) error {
+	for _, miniBlock := range body.MiniBlocks {
+		isCrossShardMiniBlockFromMe := miniBlock.SenderShardID == tc.shardCoordinator.SelfId() &&
+			miniBlock.ReceiverShardID != tc.shardCoordinator.SelfId()
+		if !isCrossShardMiniBlockFromMe {
```
Comment (raduchis) [inline process/coordinator/process.go:1222]: no need for gasCheck for senderShard on crossShard transactions?

**Review by raduchis — DISMISSED** [review]

**Review by valentincraciun99 — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Review by raduchis — APPROVED** [review]

**Review by valentincraciun99 — APPROVED** [review]
system testing

**Pull request merged**